### ISSUE TITLE
Scenario: Retry Rule Strategy

### ISSUE THREAD

**Issue by krt-c** [LightBDD/LightBDD#257]
#### Description
I would like to implement a retry rule strategy for flaky tests which are beyond our control such as intermittent time outs from a third party service which are very difficult to occur in a real environment, such issues are causing false negatives causing our stake holders  to ignore legitimate failures.  I understand that this feature might have its side effect such as hiding actual intermittent issues but the idea is to have this feature as a selective implementation on one offs scenarios not for all scenarios. 

I implemented such a feature by doing an extension function RunUsingRetryRule which extends the IScenarioRunner.RunAsync function. Anther option was to wrap a scenario in a retry mechanism rule such as Polly...

The issue I am having is mainly on reporting because the test is still being marked as failed although it passes on X amount of retries but fails once. I was thinking of adding a another status to show the amount of retried tests and only mark a test as failed once it triggers the retry rule failure criteria but am having issues on where to implement this code within lightbdd.

I would like to get another opinion about this and how would you implement a Retry Rule strategy feature in lightbdd. 

Thanks! 

https://itnext.io/how-to-retry-fails-automation-tests-af4884c47712

**Comment by Suremaker** [comment]
Hello @krt-c ,

Here is the example of xunit scenario with `RetryStep` decorator:

```c#
public class Retried_step_fixture : FeatureFixture
{
    private int _attempt = 0;

    [Scenario]
    public void Scenario_example()
    {
        Runner.RunScenario(
            _ => Given_step(),
            _ => When_step(),
            _ => Then_step());
    }

    private void Given_step()
    {
        StepExecution.Current.Comment("Given");
    }

    [RetryStep]
    private void When_step()
    {
        StepExecution.Current.Comment("When");
        ++_attempt;
        Assert.Equal(3, _attempt);
    }

    private void Then_step()
    {
        StepExecution.Current.Comment("Then");
    }
}

class RetryStepAttribute : Attribute, IStepDecoratorAttribute
{
    public int Order { get; }
    public async Task ExecuteAsync(IStep step, Func<Task> stepInvocation)
    {
        int i = 0;
        while (true)
        {
            try
            {
                await stepInvocation();
                return;
            }
            catch (ScenarioExecutionException ex)
            {
                if (++i > 3)
                    throw;
                step.Comment($"Repeating on: {ex.InnerException.GetType()} - {ex.InnerException.Message}");
            }
        }
    }
}
```

If you run it, the following output will get produced:
```
SCENARIO: Scenario example
  STEP 1/3: GIVEN step...
  STEP 1/3: /* Given */
  STEP 1/3: GIVEN step (Passed after 5ms)
  STEP 2/3: WHEN step...
  STEP 2/3: /* When */
  STEP 2/3: /* Repeating on: Xunit.Sdk.EqualException - Assert.Equal() Failure
Expected: 3
Actual:   1 */
  STEP 2/3: /* When */
  STEP 2/3: /* Repeating on: Xunit.Sdk.EqualException - Assert.Equal() Failure
Expected: 3
Actual:   2 */
  STEP 2/3: /* When */
  STEP 2/3: WHEN step (Passed after 3ms)
  STEP 3/3: THEN step...
  STEP 3/3: /* Then */
  STEP 3/3: THEN step (Passed after <1ms)
  SCENARIO RESULT: Passed after 24ms
```
where the test will be presented as passed in both xunit and lightbdd reports:
![image](https://user-images.githubusercontent.com/1894621/172470114-bd8e2dd8-1141-4b83-b616-47015431ac1d.png)
![image](https://user-images.githubusercontent.com/1894621/172470193-afdd22ee-f4b8-4b71-9d31-06b19d5f17b5.png)

Of course the `StepExecution.Current.Comment()` lines are optional, present here to visualize the behavior.

Please let me know if that approach is sufficient to overcome your issue with flakey tests.
If you need to retry the whole scenario, please let me know - I would need to make it possible on the LightBDD engine.

**Comment by krt-c** [comment]
thanks for your reply @Suremaker. we do something similar to that through Polly at step level. What we require is more at scenario level.

**Comment by Suremaker** [comment]
Hello @krt-c ,

Please take a look at this code example:
```c#
public class Retried_scenario_fixture : FeatureFixture
{
    private int _attempt = 0;

    [Scenario]
    [RetryScenario]
    public void Repeated_scenario()
    {
        Runner.RunScenario(
            _ => Given_step(),
            _ => When_step(),
            _ => Then_step());
    }

    private void Given_step()
    {
        StepExecution.Current.Comment("Given");
    }

    private void When_step()
    {
        StepExecution.Current.Comment("When");
        ++_attempt;
    }

    private void Then_step()
    {
        StepExecution.Current.Comment("Then");
        Assert.Equal(3, _attempt);
    }
}

class RetryScenarioAttribute : Attribute, IScenarioDecoratorAttribute
{
    public async Task ExecuteAsync(IScenario scenario, Func<Task> scenarioInvocation)
    {
        int i = 0;
        while(true)
        {
            try
            {
                await scenarioInvocation();
                return;
            }
            catch
            {
                if (++i > 3) 
                    throw;
            }
        }
    }

    public int Order { get; }
}
```

It does the retry on the scenario level. As of LightBDD 3.4.1 there is a bug that the repeated scenario exception details are not cleared - it is being addressed in #258 

Please let me know if that solution would be helpful for you. If so, I'll merge in the changes and release version 3.4.2

**Comment by krt-c** [comment]
Thanks @Suremaker for your update.  The issue I see is that for a retry strategy to be considered a success the test needs to pass when its a success for over 50% of the total retries OR if the first attempt is a success,

So if we are take a a total retries of 3. 

1st [fail]/ 2nd  [success] / 3rd [fail] = fail
1st [fail] / 2nd [success]  / 3rd [success] = success
2 consecutive fails = fail
1st [success] = first attempt success should be considered a success.

(ofc this can be added as a separate decorator from the one mentioned above)

**Comment by Suremaker** [comment]
Hey @krt-c ,

I have thought about your problem.
With the current scenario decorators it would be possible to achieve it (with caveats described below):

Take a look at the following code:
```c#
class RetryScenarioAttribute : Attribute, IScenarioDecoratorAttribute
{
    private readonly int _maxRetries;
    private readonly double _successRate;

    public RetryScenarioAttribute(int maxRetries, double successRate)
    {
        _maxRetries = maxRetries;
        _successRate = successRate;
    }
    public async Task ExecuteAsync(IScenario scenario, Func<Task> scenarioInvocation)
    {
        var attempts = new List<bool>();
        while (true)
        {
            try
            {
                await scenarioInvocation();

                if (attempts.Count == 0)
                    return; //first success

                attempts.Add(true);
                if (attempts.Count(a => a == true) / (double)_maxRetries >= _successRate)
                    return;
            }
            catch
            {
                attempts.Add(false);
                var currentFailureRate = attempts.Count(a => a == false) / (double)_maxRetries;
                var maxFailureRate = 1 - _successRate;
                if (currentFailureRate > maxFailureRate)
                    throw new InvalidOperationException($"Current failure rate ({currentFailureRate}) exceeded max failure rate ({maxFailureRate})");
            }
        }
    }

    public int Order { get; }
}
```

For this scenario, it will pass on first try:
```c#
public class Retried_scenario_fixture : FeatureFixture
{
    private int _attempt = 0;

    [Scenario]
    [RetryScenario(3, 0.6)] //three repeats, 0.6 success rate
    public void Repeated_scenario()
    {
        Runner.RunScenario(
            _ => Given_step(),
            _ => When_step(),
            _ => Then_step());
    }

    private void Given_step()
    {
        StepExecution.Current.Comment("Given");
    }

    private void When_step()
    {
        StepExecution.Current.Comment("When");
        ++_attempt;
    }

    private void Then_step()
    {
        StepExecution.Current.Comment("Then");
        Assert.True(_attempt % 2 == 1);
    }
}
```
```
SCENARIO: Repeated scenario
  STEP 1/3: GIVEN step...
  STEP 1/3: /* Given */
  STEP 1/3: GIVEN step (Passed after 5ms)
  STEP 2/3: WHEN step...
  STEP 2/3: /* When */
  STEP 2/3: WHEN step (Passed after <1ms)
  STEP 3/3: THEN step...
  STEP 3/3: /* Then */
  STEP 3/3: THEN step (Passed after <1ms)
  SCENARIO RESULT: Passed after 20ms
```

For this one, it will pass on third attempt:
```c#
public class Retried_scenario_fixture : FeatureFixture
{
    private int _attempt = 0;

    [Scenario]
    [RetryScenario(3, 0.6)]
    public void Repeated_scenario()
    {
        Runner.RunScenario(
            _ => Given_step(),
            _ => When_step(),
            _ => Then_step());
    }

    private void Given_step()
    {
        StepExecution.Current.Comment("Given");
    }

    private void When_step()
    {
        StepExecution.Current.Comment("When");
        ++_attempt;
    }

    private void Then_step()
    {
        StepExecution.Current.Comment("Then");
        Assert.True(_attempt > 1); //this is a different assert
    }
}
```

```
SCENARIO: Repeated scenario
  STEP 1/3: GIVEN step...
  STEP 1/3: /* Given */
  STEP 1/3: GIVEN step (Passed after 5ms)
  STEP 2/3: WHEN step...
  STEP 2/3: /* When */
  STEP 2/3: WHEN step (Passed after <1ms)
  STEP 3/3: THEN step...
  STEP 3/3: /* Then */
  STEP 3/3: THEN step (Failed after 18ms)
  STEP 1/3: GIVEN step...
  STEP 1/3: /* Given */
  STEP 1/3: GIVEN step (Passed after <1ms)
  STEP 2/3: WHEN step...
  STEP 2/3: /* When */
  STEP 2/3: WHEN step (Passed after <1ms)
  STEP 3/3: THEN step...
  STEP 3/3: /* Then */
  STEP 3/3: THEN step (Passed after <1ms)
  STEP 1/3: GIVEN step...
  STEP 1/3: /* Given */
  STEP 1/3: GIVEN step (Passed after <1ms)
  STEP 2/3: WHEN step...
  STEP 2/3: /* When */
  STEP 2/3: WHEN step (Passed after <1ms)
  STEP 3/3: THEN step...
  STEP 3/3: /* Then */
  STEP 3/3: THEN step (Passed after <1ms)
  SCENARIO RESULT: Passed after 39ms
```

And for this one, it will fail!
```c#
public class Retried_scenario_fixture : FeatureFixture
{
    private int _attempt = 0;

    [Scenario]
    [RetryScenario(3, 0.6)]
    public void Repeated_scenario()
    {
        Runner.RunScenario(
            _ => Given_step(),
            _ => When_step(),
            _ => Then_step());
    }

    private void Given_step()
    {
        StepExecution.Current.Comment("Given");
    }

    private void When_step()
    {
        StepExecution.Current.Comment("When");
        ++_attempt;
    }

    private void Then_step()
    {
        StepExecution.Current.Comment("Then");
        Assert.True(_attempt % 2 == 0); //different assert
    }
}
```

```
SCENARIO: Repeated scenario
  STEP 1/3: GIVEN step...
  STEP 1/3: /* Given */
  STEP 1/3: GIVEN step (Passed after 5ms)
  STEP 2/3: WHEN step...
  STEP 2/3: /* When */
  STEP 2/3: WHEN step (Passed after <1ms)
  STEP 3/3: THEN step...
  STEP 3/3: /* Then */
  STEP 3/3: THEN step (Failed after 19ms)
  STEP 1/3: GIVEN step...
  STEP 1/3: /* Given */
  STEP 1/3: GIVEN step (Passed after <1ms)
  STEP 2/3: WHEN step...
  STEP 2/3: /* When */
  STEP 2/3: WHEN step (Passed after <1ms)
  STEP 3/3: THEN step...
  STEP 3/3: /* Then */
  STEP 3/3: THEN step (Passed after <1ms)
  STEP 1/3: GIVEN step...
  STEP 1/3: /* Given */
  STEP 1/3: GIVEN step (Passed after <1ms)
  STEP 2/3: WHEN step...
  STEP 2/3: /* When */
  STEP 2/3: WHEN step (Passed after <1ms)
  STEP 3/3: THEN step...
  STEP 3/3: /* Then */
  STEP 3/3: THEN step (Failed after <1ms)
  SCENARIO RESULT: Failed after 41ms
    Scenario: System.InvalidOperationException : Current failure rate (0,6666666666666666) exceeded max failure rate (0,4)
    	at LightBDD_Example.RetryScenarioAttribute.ExecuteAsync(IScenario scenario, Func`1 scenarioInvocation) in c:\tmp\lightbdd_example\LightBDD_Example\LightBDD_Example\Retried_scenario_fixture.cs:line 79
    Step 3: Xunit.Sdk.TrueException : Assert.True() Failure
    	Expected: True
    	Actual:   False
    	at LightBDD_Example.Retried_scenario_fixture.Then_step() in c:\tmp\lightbdd_example\LightBDD_Example\LightBDD_Example\Retried_scenario_fixture.cs:line 43
    	at lambda_method31(Closure , NoContext , Object[] )
```

### Caveats

1. Currently it won't be possible to see the details of the previous runs, nor the number of the attempts. The previous runs are discarded. They will however add up to the overall scenario execution time.
2. The retry operation will be performed on the same instance of the test class (nunit, xunit and others frameworks) so the field state produced by previous runs will leak to subsequent attempts. It is unfortunately true even for the [contextual scenarios](https://github.com/LightBDD/LightBDD/wiki/Scenario-Steps-Definition#contextual-scenarios) as the scenario decorator is called after the scenario context is created. It could be mitigated by wrapping the scenario steps withing additional composite step with own context, but that would make the scenarios less readable (I can elaborate on it though if you would like to go this route).

---
To overcome these caveats I would need to make much more significant changes in the LightBDD engine, which I would be happy to look into, however it will take time.

**Comment by Suremaker** [comment]
Let me know what you think about it - regardless, I will merge and release the small bug fix that I have put in the PR.

**Comment by krt-c** [comment]
hi @Suremaker,  sorry for the late reply was away the last few days. I understand I think I can make things work with the examples you provided above. ofc it would be nice to address the mentioned limitations since am worried that results from the previous runs might interfere with the subsequent attempts.  

Thanks again for looking into this! much appreciated!

**Comment by Suremaker** [comment]
Thanks @krt-c .

As I have implemented a fix #257 and provided the details on how to implement a retry strategy using the current functionality, I will close this particular issue.

I have created #259 to provide a better support for these scenarios.

**Comment by kyurkchyan** [comment]
@Suremaker 

Thanks a lot for sharing the implementation details. Unfortunately the retry scenario attribute doesn't work for scenarios that are theories with inline data 

```csharp
[Scenario]
    [RetryScenario]
    [Category(TestCategories.ReadOnlyTests)]
    [InlineData(SalesAgentStatus.Active)]
    [InlineData(SalesAgentStatus.Closed)]
    public Task SearchAgents_ShouldShowOnlyAgentsMatchingTheStatusFilter(SalesAgentStatus status)
    {
        return Runner
               .AddAsyncSteps(_ => GivenTheUserIsOnTheSearchAgentsPage(),
                              _ => WhenTheyApplyAStatusFilter(status),
                              _ => ThenOnlySalesAgentsWithTheStatusShouldBeShown(status))
               .RunAsync();
    }
```

I have the following setup. And retry attribute is not applied. I guess we should use a different structure when working with inline data?
